Implementing retries on failed user updates when finishing a quest

Hus274 opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8035 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Added a finite retry system for individual user updates when finishing a quest. This should help prevent users not getting the quest rewards when their party finishes a quest. The retry system currently allows up to 30 seconds of retries but this can be adjusted as needed depending on the load and timeouts seen by mongo.

Let me know if you have any comments!

Hus274 committed:
Implementing retries on failed user updates when finishing a quest. fixes #8035

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -623,11 +623,45 @@ schema.methods.finishQuest = async function finishQuest (quest) {
 
-  return await User.update(q, updates, {multi: true}).exec();
+  if (this._id === TAVERN_ID) {
+    return await User.update(participants, updates, {multi: true}).exec();
```
If we're going to return early here for tavern quests, we don't need the ternary above.

crookedneighbor reviewed (changes requested):
I don't see any tests around the retry logic. 

This whole thing makes me nervous. To be honest, if there's an error when saving the user, I'd prefer to be notified avout the error so we can actually figure out what is casusing this. 

If we must do retries, I'd prefer them to be a set number, rather than a time period.

Hus274 commented:
> I don't see any tests around the retry logic.

I did some manual testing and it worked as expected. I agree the behavior should have automated tests around it but I'm not familiar with a way to do that here. Would there be a way to mock out the mongo call and have it return an error on the first call and succeed the second time?

> This whole thing makes me nervous. To be honest, if there's an error when saving the user, I'd prefer to be notified about the error so we can actually figure out what is causing this.

From my understanding of the problem described in the issue, it sounds like it's believed that this is due to load and timeouts on the database which is generally handled with retries and spacing the calls out. Is there a way to log the error if it happens the first time so it can be looked up? Otherwise, if we don't want to do retries, I can do what @Alys suggested and simply write it to the groups object.

>If we must do retries, I'd prefer them to be a set number, rather than a time period.

Yeah that makes sense to me. I used time cause it allows you to control the total time you're willing to spend retrying across any number of failed users but I'm ok with switching this over.

Hus274 committed:
Refactoring mongo db retries to use the same as code path as original call and moving retries to count based over time based.

Hus274 commented:
I refactored this to have the retries use the same code path as the original mongoose call meaning that the retry logic code path is now tested. I've also updated the retries to be count based instead of time based. Let me know your thoughts now with these changes.

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -623,11 +642,20 @@ schema.methods.finishQuest = async function finishQuest (quest) {
   });
 
-  let q = this._id === TAVERN_ID ? {} : {_id: {$in: this.getParticipatingQuestMembers()}};
+  let participants = this.getParticipatingQuestMembers();
```
This variable should be after the tavern logic.

Hus274 replied:
The problem is that participants depends on this.quest not being empty which happens on the following line and before the tavern logic. Because this.quest = {}; needs to be before the tavern logic and participants needs to be before that, I think this needs to remain above the tavern logic. Let me know if you have other thoughts here or if I'm misinterpreting something.

crookedneighbor replied:
Hmm. I think this will not work with tavern quests then. I'd have to look at the getParticipatingQuestMembers method again, because I'm not sure how it interacts with a tavern quest.

Hus274 replied:
That's why I left the ternary operator in the beginning because I wasn't sure of the impact of calling this with a tavern quest. The getParticipatingMembers method depends on this.quest.members which I'm not sure the state of for tavern quests as I don't know how they are started or populated.

I can add the ternary operation back or make the getParticipatingQuestMembers method explicitly handle tavern quests to return an empty object. What do you think?

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -623,11 +642,20 @@ schema.methods.finishQuest = async function finishQuest (quest) {
+  }
+
+  let promises = [];
+  _.each(participants, userId => {
```
I don't remember, is participants an array? If so, why not use map?

```js
let promises = participants.map(userId => _updateUserWithRetries(userId, updates))

Hus274 replied:
Makes sense, I don't know why I never think to use the map function in javascript. I've updated this.

crookedneighbor commented on test/api/v3/unit/models/group.test.js:
```diff
@@ -1173,11 +1173,12 @@ describe('Group Model', () => {
         it('updates participating members with rewards', async () => {
           await party.finishQuest(quest);
 
-          expect(User.update).to.be.calledOnce;
```
There's still no test for the retry logic. You can use the sinon spy library to provide the results you need for failures to update.

Hus274 replied:
Makes sense, thanks for the pointer! I've updated this to include tests with no failures, test with one failure followed by success, and then test for failing infinitely. Let me know if there are any other tests you think I should add.

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -577,6 +577,25 @@ schema.statics.cleanGroupQuest = function cleanGroupQuest () {
+    saved: false,
+    retries: 0,
+  };
+  while (!data.saved && data.retries < 5) {
```
Extract 5 into a variable like 

```js 
const MAX_UPDATE_RETRIES = 5;
``` 

So we don't have a magic number.

Hus274 replied:
Good call, updated.

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -577,6 +577,25 @@ schema.statics.cleanGroupQuest = function cleanGroupQuest () {
+    retries: 0,
+  };
+  while (!data.saved && data.retries < 5) {
+    await User.update({_id: userId}, updates).exec((err, raw) => { // eslint-disable-line babel/no-await-in-loop
```
I'm not confident that this will even work in a loop, to be honest.

Hus274 replied:
I agree this was kind of messy and made the scopes hard to follow. I've updated this to a recursive method that I think is much cleaner.

Hus274 committed:
Adding tests for retry logic and updating retries to happen recursively

Hus274 commented:
I've updated this according to your review. I've added tests around the retry logic and moved the retry logic to a recursive method as it seems to be a good bit cleaner. Let me know any more feedback you have

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -650,9 +644,8 @@ schema.methods.finishQuest = async function finishQuest (quest) {
-  let promises = [];
-  _.each(participants, userId => {
-    promises.push(_updateUserWithRetries(userId, updates));
+  let promises = _.map(participants, userId => {
```
Is there a reason to use lodash over the native implementation of map?

Hus274 replied:
Nope, I just thought the standard was to use lodash functions. My mistake, I've fixed this.

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -577,6 +578,18 @@ schema.statics.cleanGroupQuest = function cleanGroupQuest () {
 };
 
+async function _updateUserWithRetries (userId, updates, numTry = 1) {
+  return User.update({_id: userId}, updates).exec((err, raw) => {
```
Is there a reason to use the callback syntax instead of the promise syntax? If I'm reading this correctly, returning these values isn't actually doing anything.

Hus274 replied:
Sorry, this is just due to my lack of experience with Javascript. I've adjusted it to the promise syntax now and it's way better. I'm not using the .catch() method because it seems it doesn't work well with mongoose's promise library and so instead I'm using the .then(function(success), function(error)) syntax. Let me know if that works for you.

Hus274 replied:
http://stackoverflow.com/questions/30672265/why-can-i-not-chain-catch-when-calling-mongoose-model-create-in-node

Hus274 commented on website/server/models/group.js:
```diff
@@ -623,11 +636,19 @@ schema.methods.finishQuest = async function finishQuest (quest) {
+    return await User.update({}, updates, {multi: true}).exec();
+  }
+
+  let promises = _.map(participants, userId => {
```
Nope, I just thought the standard was to use lodash functions. My mistake, I've fixed this.

Hus274 committed:
Moving callbacks to promises and other tweaks according to pr.

Hus274 commented:
I moved the getParticipatingQuestMembers back to a ternary operator for now and addressed your other comments. I can definitely re-remove the ternary operator if you want, just let me know. Thanks for taking all the time to review this, I appreciate it.

paglias commented on website/server/models/group.js:
```diff
@@ -577,6 +578,19 @@ schema.statics.cleanGroupQuest = function cleanGroupQuest () {
 
+async function _updateUserWithRetries (userId, updates, numTry = 1) {
+  return await User.update({_id: userId}, updates).exec()
+    .then((raw) => {
```
you should be able to use catch, it's used in multiple places with mongoose. Because we're plugging Bluebird in place of their promise implementation that doesn't support it

Hus274 replied:
Ahh ok that makes sense. I'll update it, thanks!

Hus274 committed:
Chaging mongoose promise to use .catch() functionality

Hus274 commented:
Updated this to use .catch() functionality, thanks @paglias!

paglias commented on website/server/models/group.js:
```diff
@@ -577,6 +578,19 @@ schema.statics.cleanGroupQuest = function cleanGroupQuest () {
+      if (numTry < MAX_UPDATE_RETRIES) {
+        return _updateUserWithRetries(userId, updates, ++numTry);
+      } else {
+        return err;
```
here you should throw err, not return it, otherwise it'll be treated as a success return value

Hus274 committed:
If all retries fail, the system will now throw an error instead of returning an error message.

Hus274 commented:
Updated to throw an error instead of returning it.
